[primTorch] decompositions for upsample_bicubic2d

fdrocha opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85403
* #85402

FYI, this decomposition seems to be significantly slower than the lowering in torchinductor:

```
------------------------------------- upsample_bicubic2d -------------------------------------]
                                                              |  lowering  |  Inductor  |  Eager
32 threads: ------------------------------------------------------------------------------------
      (torch.Size([16, 4, 128, 256]),), ((512, 1024), True)   |    1.8     |   3.880    |   1.4
      (torch.Size([16, 4, 128, 256]),), ((512, 1024), False)  |    1.9     |   3.887    |   1.4
```

This seems related to the fact that in the lowering we can use int32s as the indices and in the decomp we can only use int64s (see https://github.com/pytorch/torchdynamo/issues/1293).

Commit:
[primTorch] decompositions for upsample_bicubic2d

[ghstack-poisoned]

lezcano commented on torch/_decomp/decompositions.py:
```diff
@@ -2030,3 +2033,89 @@ def binary_cross_entropy_with_logits(
+    return _upsample_cubic_interp1d(coeffs_y, t_y)
+
+
+@register_decomposition(aten.upsample_bicubic2d.vec)
+@out_wrapper()
+@pw_cast_for_opmath
+def upsample_bicubic2d_vec(
+    a: Tensor,
+    output_size: Optional[Tuple[int, int]],
+    align_corners: bool,
+    scale_factors: Optional[Tuple[float, float]] = None,
+) -> Tensor:
+    def compute_output_size(input_size, output_size=None, scale_factors=None):
+        utils.check(
+            bool(output_size) + bool(scale_factors) == 1,
+            lambda: "Must specify exactly one of output_size and scale_factors.",
+        )
+        if output_size:
+            return output_size
+        else:
+            return tuple(
+                int(w * scale) for w, scale in zip(input_size[2:], scale_factors)
+            )
+
+    output_size = compute_output_size(a.shape, output_size, scale_factors)
+    scale_h = scale_factors[0] if scale_factors else None
+    scale_w = scale_factors[1] if scale_factors else None
+    return upsample_bicubic2d_default(a, output_size, align_corners, scale_h, scale_w)
```
It's not clear why this decomposition is necessary, given that this function is `CompositeExplicit`.

fdrocha replied:
I just filled an issue pertaining to this, see https://github.com/pytorch/pytorch/issues/85405

Commit:
Update on "[primTorch] decompositions for upsample_bicubic2d"


FYI, this decomposition seems to be significantly slower than the lowering in torchinductor:

```
------------------------------------- upsample_bicubic2d -------------------------------------]
                                                              |  lowering  |  Inductor  |  Eager
32 threads: ------------------------------------------------------------------------------------
      (torch.Size([16, 4, 128, 256]),), ((512, 1024), True)   |    1.8     |   3.880    |   1.4
      (torch.Size([16, 4, 128, 256]),), ((512, 1024), False)  |    1.9     |   3.887    |   1.4
```

This seems related to the fact that in the lowering we can use int32s as the indices and in the decomp we can only use int64s (see https://github.com/pytorch/torchdynamo/issues/1293).


[ghstack-poisoned]

Commit:
Update on "[primTorch] decompositions for upsample_bicubic2d"


FYI, this decomposition seems to be significantly slower than the lowering in torchinductor:

```
------------------------------------- upsample_bicubic2d -------------------------------------]
                                                              |  lowering  |  Inductor  |  Eager
32 threads: ------------------------------------------------------------------------------------
      (torch.Size([16, 4, 128, 256]),), ((512, 1024), True)   |    1.8     |   3.880    |   1.4
      (torch.Size([16, 4, 128, 256]),), ((512, 1024), False)  |    1.9     |   3.887    |   1.4
```

This seems related to the fact that in the lowering we can use int32s as the indices and in the decomp we can only use int64s (see https://github.com/pytorch/torchdynamo/issues/1293).


[ghstack-poisoned]

Commit:
Update on "[primTorch] decompositions for upsample_bicubic2d"


FYI, this decomposition seems to be significantly slower than the lowering in torchinductor:

```
------------------------------------- upsample_bicubic2d -------------------------------------]
                                                              |  lowering  |  Inductor  |  Eager
32 threads: ------------------------------------------------------------------------------------
      (torch.Size([16, 4, 128, 256]),), ((512, 1024), True)   |    1.8     |   3.880    |   1.4
      (torch.Size([16, 4, 128, 256]),), ((512, 1024), False)  |    1.9     |   3.887    |   1.4
```

This seems related to the fact that in the lowering we can use int32s as the indices and in the decomp we can only use int64s (see https://github.com/pytorch/torchdynamo/issues/1293).


[ghstack-poisoned]

fdrocha commented:
@ngimel, @ezyang could you PTAL? Mostly this was discussed in other PRs and issues

ngimel reviewed (approved):
lgtm, thanks

lezcano commented:
@pytorchbot merge
